Added new optional fields for the `parameters` dict that is used to initialise the `GridScan` class

LawrenceDior opened the pull request:
The new fields are as follows:
- `"n_samples"` - this overrides the `n_samples` argument passed to the `GridScan` constructor, potentially resulting in a sampling grid that has dimensions of varying sizes.
- `"endpoint"` - if `False`, the maximum value of the range specified for this parameter is excluded. This may be convenient for parameters that represent angles.
- `"values"` - if this field is present, it is used to set the parameter values directly, and all other fields for this parameter are ignored. This gives the user full control over the values to include in the parameter scan.

ZedThree commented:
Lovely, thanks @LawrenceDior! Just one question really, how does/should `values` interact with `n_samples` when `len(values) > n_samples`? I think there's several different choices you might want: take all of them (as is currently the case here), take the first `n_samples`, or take a random set. I think the choice you've made is fine, but it's probably best to document this

LawrenceDior commented:
> Lovely, thanks @LawrenceDior! Just one question really, how does/should `values` interact with `n_samples` when `len(values) > n_samples`? I think there's several different choices you might want: take all of them (as is currently the case here), take the first `n_samples`, or take a random set. I think the choice you've made is fine, but it's probably best to document this

I think `len(values)` should override `n_samples`, just as the new optional `n_samples` field of the `parameters` dict overrides the general `n_samples` argument of the `GridScan` constructor if the former is provided for a given parameter. I've updated the documentation to make this clearer.

LawrenceDior commented:
@ZedThree, as it stands, `values` is not accounted for in the `LatinHypercubeSampler` class, but I've had some thoughts about a few different ways it could work and was wondering if you have a preference:

1. Leave it out and require the user to specify a continuous range for every parameter if they plan to use this sampling method.
2. Sample over a continuous range from `min(values)` to `max(values)`, and this can be linearly-spaced by default or log-spaced if the user has set the `"log"` field to `True`.
3. Treat `values` as a strictly discrete set of values to sample from. In the loop over `self._parameters.items()` in the `LatinHypercubeSampler` constructor, if `"values"` is present, the `"min"` and `"max"` fields are set to `-0.5` and `len(values) - 0.5` respectively, and `"log"` is set to `False`. Samples within this continuous range can then be rounded to the nearest integer and used as indices into `values`. Note that this will result in some repeat values if the sample size exceeds `len(values)`.

LawrenceDior commented:
Hi @JoelLucaAdams, thanks for the feedback. I've added a UserWarning that's set off by any other keys that appear alongside `"values"` for a given parameter. I've also added a docstring to `LatinHypercubeSampler` which clarifies the differences between the ways the `GridScan` and `LatinHypercubeSampler` classes handle their respective `parameters` arguments.

JoelLucaAdams commented:
@LawrenceDior Looks great, thanks very much! I consider this to be adding plentiful functionality so please add your name to the `CITATION.cff` and `pyproject.toml` and I'll approve the PR :)
